_episodes/18.md

```diff
@@ -0,0 +1,560 @@
+   nationkey bigint,
+   phone varchar(15),
+   acctbal double,
+   mktsegment varchar(10),
+   comment varchar(117)
+)
+```
+
+
+
+
+### Views and materialized views:
+
+The structure of a view is similar to tables in that they have columns, rows, 
+and schemas similar to regular database tables. What then do views offer over
+tables? Views offer ways to encapsulate complex SQL statements. For instance, 
+take this SQL query that would run over the customer and orders tables defined 
+before.
+
+```
+SELECT 
+ c.custkey, 
+ name, 
+ nationkey, 
+ mktsegment, 
+ sumtotalprice, 
+ openstatuscount, 
+ failedstatuscount, 
+ partialstatuscount
+FROM 
+ customer c 
+ JOIN (
+  SELECT 
+   custkey, 
+   SUM(totalprice) AS sumtotalprice, 
+   COUNT_IF(orderstatus = 'O') AS openstatuscount,
+   COUNT_IF(orderstatus = 'F') AS failedstatuscount, 
+   COUNT_IF(orderstatus = 'P') AS partialstatuscount
+  FROM orders
+  GROUP BY custkey
+ ) o
+ ON c.custkey = o.custkey;
+```
+
+This query performs some aggregations on the orders table grouped by customer.
+Then there is a join performed on the aggregated orders table and customer by
+customer.
+
+|custkey|name|nationkey|mktsegment|sumtotalprice|openstatuscount|failedstatuscount|partialstatuscount|
+|-------|----|---------|----------|-------------|---------------|-----------------|------------------|
+|376|Customer#000000376|16|AUTOMOBILE|1600696.4700000002|3|6|1|
+|377|Customer#000000377|23|MACHINERY|803271.9400000001|3|6|0|
+|379|Customer#000000379|7|AUTOMOBILE|3155009.54|7|11|0|
+
+
+From here, there are many ways you could further evaluate the resulting data. 
+You could filter and look at which market segment is spending the most on your
+products. You could also look at where there are the most failed orders by the
+nation column to evaluate where shipping lines may need to be improved. The 
+table above which results from the example query is a good intermediate state of 
+the data that can be reused for many future evaluations. Instead of defining a 
+new table, you can create a view on this data that will encapsulate the complex
+SQL that was used to calculate it. This is done using the `CREATE VIEW` 
+statement.
+
+```
+CREATE VIEW customer_orders_view AS 
+<complex SQL query above>
+```
+
+Now, when you want to run any further analysis on this intermediate dataset, you
+simply refer to the view instead of having to rewrite the statement before. As
+mentioned, this view also has a schema and is treated much like a table when the
+query engine does its planning. In this way it is also easier to map the data to
+the application logic by enabling different shapes of the same data.
+
+Another reason why you would want to create a view is to control access to the
+data. When running the query, you get to choose which columns and rows get
+filtered out and that return from when users query the view. The authorization 
+of a user is only to read the abstraction of the data rather than the real data.
```

Comment:
of a use is tied to the view and its content, and that can significantly differ from the complete data in the underlying tables. For example, the views can exclude sensitive data like social security numbers, birth dates, credit card numbers, and many other facts.